### DISCUSSION TITLE
Interactive map v0.0

### DISCUSSION THREAD

**Discussion by aplaice** [axelboc/anki-ultimate-geography#649] [General]
To those not aware, @helitopia [has been doing amazing work](https://github.com/anki-geo/ultimate-geography/pull/644), developing an interactive map
to enhance the "Country - Map" cards.

The interactive map allows you to select the given country/region, on the "front" side, and displays the correct one, with colour-coding telling you whether you answered right or wrong (currently green for right, red for wrong).

https://github.com/user-attachments/assets/b0b9c771-7f72-404c-9e92-a021bdf3598c

https://github.com/user-attachments/assets/666f2e35-d31a-45fa-9a8f-e08487e78e5a

The implementation is fully local, with all the map data and javascript code on your device(s), and no external calls to any other services. This means that it can work off-line, respects your privacy, is secure (in the sense that nobody can replace the version you have with a malicious one) and is reliable (its continued functioning is not dependent on, say, any first- or third-party servers)!  All the new data and code is just 1.3 MB (less than a typical photo), so it also doesn't take up much storage space on your system.

<hr/>

**We would love it if you tried the new interactive map out, so that you could tell us what you think!** (What you like/what annoys you/what would make the map better?) We'd also like to know if the system works on various systems (**especially iOS ­— iPhone/iPad** — to which we doesn't have easy access)!

Note that the current implementation is still a beta, with known issues, most due to current[^1] limitations in the underlying jsvectormap library[^2]. Most importantly, the interactive map currently does not work for all regions, especially small dependent territories and seas — the interactive map _is_ displayed on the front side (to help visualisation), but on the back side, the static map is shown!

[^1] We intend to extend the library either upstream or via a fork to remove these limitations.

[^2] This is not intended as criticism of the library — it's very lightweight and convenient, just not yet fully adapted to our needs.

<hr/>

In the future, the deck with the interactive map is intended as an upgrade for the current extended deck (such that you can keep your review history). However, for now, **purely for testing purposes**, we've created a test deck, which can be imported in parallel to the "normal" extended/standard AUG decks. Reviewing in this deck will not affect your "normal" AUG deck. You can also safely delete this "Ultimate Geography experimental interactive" deck, after testing, without losing your notes/cards/reviews in the "normal" deck.

(Alternatively you can just import into a separate Anki profile.)

If you want a deck with just the interactive "Country - Map" cards, you can create a filtered deck, with this search: `"deck:Ultimate Geography experimental interactive" "card:Country - Map"`.  (If you create it, we recommend emptying this filtered deck before deleting the actual "Ultimate Geography experimental interactive" deck.)

The import procedure is the same as for the "normal" decks:

https://github.com/anki-geo/ultimate-geography/blob/master/README.md#getting-started

except that you need to use the zip file below!

[Ultimate_Geography_v0.0_EN_EXPERIMENTAL.zip](https://github.com/user-attachments/files/16657197/Ultimate_Geography_v0.0_EN_EXPERIMENTAL.zip)

(We can also provide zip files for other languages if these are desired!)

(If you do want to upgrade, the relevant zip files are [here](https://github.com/aplaice/anki-ultimate-geography/releases/tag/everfresh). If you're upgrading from the Extended deck, then you just need to import the deck, as normal, and approve the several pop-up messages. A general guide for managing the pop-ups when switching deck type is available [here](https://github.com/anki-geo/ultimate-geography/wiki/Switching-deck-type). Note that undoing the upgrade — switching back from the Interactive deck to the Extended deck — should also be a smooth process, as described in the aforementioned guide.)

An introduction to the way the experiment deck and interactive maps work is [here](https://github.com/anki-geo/ultimate-geography/wiki/Experimental-extended-deck).

<hr/>

Thanks again to @helitopia for his fantastic idea and his meticulous implementation!

**Comment by josealberto4444** [comment]
Wow, this is fantastic! I love it! 🥹

When I review on Anki desktop, I have to click on the country and then on Show answer. If I don't click on Show answer, the country remains selected but nothing else happens. Is this expected or maybe due to some config I did and I don't remember? =P

I'll take a look at AnkiMobile soon on an iPhone and an iPad and will let you know. ^^

**Reply by helitopia** [reply]
Hi,

Could you please:
- when reviewing one of the cards with interactive map press "E" -> select "Cards" -> and make a screenshot of text inside `<script>` element (Front Template)
  For example:
  <img src="https://github.com/user-attachments/assets/f0181a5f-2fbc-4f1e-9018-412305f65c40" height="100px"/>
  This way we will be able to see the current configuration including whether the option to automatically swap to answer side is enabled.
- specify your current Anki version on which the issue is reproduced

**Reply by josealberto4444** [reply]
The screenshot:
![Screenshot_20240822_085738](https://github.com/user-attachments/assets/c4202ef6-8dba-42c0-9037-af782966b184)
The same text copied:
```
<script>
  sessionStorage.setItem("userConfig",
    JSON.stringify({
      commonFeatures: {
        interactiveEnabled: "true",
        interactiveMobileEnabled: "true",
        autoAnswerEnabled: "true",
        toolTipEnabled: "false",
      },
      commonColors: {
        region: "#fdfbe5",
        selectedRegion: "#e7f3ea",
        incorrectRegionHighlight: "#c02637",
        correctRegionHighlight: "#329446",
        background: "#b3dff5",
        border: "#757674",
        tooltipBackground: "#fdfbe5",
        tooltipText: "#000000"
      }
    }));
</script>
```

The Anki version:
Version ⁨24.06.3 (d678e393)⁩
Python 3.12.4 Qt 6.7.2 PyQt 6.7.1

Thank you very much! I have a couple more reports/issues tested, should I report them here or in the repo issues? ^^ @aplaice @axelboc @helitopia

**Reply by axelboc** [reply]
In this thread please, as new comments. We can hide individual comment threads as they're resolved.

**Reply by helitopia** [reply]
Hello @josealberto4444 

Will have a look at it asap, as everything with configuration seems normal and the map is supposed to automatically swap to back card side.

**Reply by helitopia** [reply]
Have looked into the issue. As I mentioned in the wiki:
> swapToBackSide() function that performs automatic swap to answer card side heavily relies on Anki APIs support for which may be removed at any moment. Thus, such feature might not be considered reliable and in the worst case scenario its existence might be deprecated while interactive map still working as usual, simply without it.

We use really fragile way to swap card to answer side - sending a [`KeyboardEvent`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent) with `code` property set to `Enter`. When writing the logic I have been using older Anki version (23.12.1).

Apparently, since version [24.06](https://github.com/ankitects/anki/releases/tag/24.06), Anki uses property `key` instead of `code` for `KeyboardEvent` events ([commit](https://github.com/ankitects/anki/commit/573af9cd4c09603f8ebf2afbfb44f9336007cf19) - [reason for the change](https://github.com/ankitects/anki/issues/3069#issue-2178272168)).

This is indeed a bug from our side (and it is really hard to blame anyone else as we accept the risk of Anki API changing as it is not intended to be even used by users).

I tried to fix it by setting both properties (for it to work on both pre-24.06 and post- versions): `new KeyboardEvent("keypress", {code: "Enter", key: "Enter"})`. It worked, so I will open PR to resolve the issue (@aplaice @axelboc or should I open issue first?).

**Reply by axelboc** [reply]
PR directly, all good 😉

**Reply by helitopia** [reply]
[The PR](https://github.com/anki-geo/ultimate-geography/pull/651)

**Comment by josealberto4444** [comment]
With respect to colours of correct and incorrect answers, currently it doesn't seem intuitive for me. When the solution is incorrect, I would expect the solution I tapped to be filled with the incorrect colour/pattern and the correct solution to be always marked with the correct colour. I only realized now, when answering.

**Reply by helitopia** [reply]
That is exactly what we [discussed early on](https://github.com/anki-geo/ultimate-geography/issues/638#issuecomment-2228181761)!
> # Idea (# 7 - Green/Red answer)
> ...
> Although ideally in case of incorrect answer two regions should be highlighted - selected in red and correct in green. But, again, the library right now lacks the ability to highlight regions with different colors (see [Note #5 section](https://github.com/anki-geo/ultimate-geography/issues/638)).

@axelboc @aplaice shall I add this to our agenda to investigate the possibility of (somewhere after "Selected region pattern highlight" item is resolved)?

**Reply by aplaice** [reply]
I think so, yes. We might want to apply both a pattern and a different colour to the second highlighted region, so placing it later on the list makes sense!

**Reply by josealberto4444** [reply]
> That is exactly what we [discussed early on](https://github.com/anki-geo/ultimate-geography/issues/638#issuecomment-2228181761)!

Oops, sorry! I had read all that issue, but there is so much information that I forgot. =P

**Reply by helitopia** [reply]
Added to [the comment](https://github.com/anki-geo/ultimate-geography/issues/638#issuecomment-2303910627):
> _Edit:_ Incorrect selected and correct regions highlight

**Comment by josealberto4444** [comment]
I think it would be a great improvement (more intuitive) if the animation didn't start from completely zoomed out, but from the position of the selection I made. Then it could be seen as "This is not the solution (marked with incorrect colour) and I'll take you to the correct solution). Is it too much work to store zooming and position from the front view of the card and then using it as the start point for the back view? I don't know much about the code of this project (or Anki, or javascript at all). 🥰

**Reply by helitopia** [reply]
This was also [mentioned](https://github.com/anki-geo/ultimate-geography/pull/644#pullrequestreview-2230889144) by @aplaice as a minor annoyance:
> The zoom resets (i.e. you abruptly zoom out and then zoom in) when switching to the back card.

**Reply by josealberto4444** [reply]
For me it's not annoying, it's confusing, and I think it'd be for new users, too. Anyway, I'll take care not to repeat things already discussed. =)

**Reply by axelboc** [reply]
> Anyway, I'll take care not to repeat things already discussed. =)

Don't worry about it, that's what a feedback thread is for! :wink: It's good to synthesize the points that were brought up in the PR but not yet addressed, as not everyone would have followed the initial development.

**Reply by helitopia** [reply]
Added to [the comment](https://github.com/anki-geo/ultimate-geography/issues/638#issuecomment-2303910627):
> _Edit:_ Zoom animation abrupt reset fixt

**Comment by josealberto4444** [comment]
Testing on Ankimobile (24.08, I think) on an iPad (iPadOS 17.4.1), every region gets colored as incorrect, no matter what you select in the front card.

**Reply by helitopia** [reply]
Thanks for testing it on IOS.

To be honest, I am not even sure how to debug the issue without buying AnkiMobile to install it even on virtual machine. I might only create Apple account and hope someone from the community would be willing to share the app via Apple Family Sharing.

**Reply by ukanuk** [reply]
Maybe you could test with the Safari browser? I think that would show the same behavior as AnkiMobile, as I believe the same WebKit engine is used across both of those.

**Reply by helitopia** [reply]
> Maybe you could test with the Safari browser? I think that would show the same behavior as AnkiMobile, as I believe the same WebKit engine is used across both of those.

What do you mean by "testing in Safari browser"? Opening and trying out interactive map in it?

Edit: I suspect the specifics of`sessionStorage` are different in AnkiMobile and the region code of a country cannot be retrieved:
```JavaScript
return commonConfig.regionCode === sessionStorage.getItem(commonConfig.selectedRegionSessionKey)
      ? commonColors.correctRegionHighlight
      : commonColors.incorrectRegionHighlight;
```
Thus `commonConfig.regionCode === undefined` and `commonColors.incorrectRegionHighlight` is displayed

**Reply by ukanuk** [reply]
Yes, I mean testing the cards with AnkiWeb in a Safari browser is likely similar in many ways to testing with AnkiMobile on iOS.

**Comment by axelboc** [comment]
I mentioned I wanted to play around with the **animated zoom on the back card**.

After setting up the `jsvectormaps` project locally (not trivial...), I was able to get the playground running and tweak the `setFocus` function to my liking so the computation basically follows a logarithmic scale:

```js
const newScale = Math.min(this._width / (bbox.width), this._height / (bbox.height));
const adjustedScale = Math.min(newScale, Math.log(newScale * 10)); // `min` so we don't zoom too much on big countries
    
return this._setScale(adjustedScale, ...)
```

This ensures that we zoom "close" to large countries and "far" from small countries. I think it's a good compromise without going into complicated solutions like showing circles or zoom boxes.

![zoom-2](https://github.com/user-attachments/assets/ad54b301-b3e3-4c7a-9868-e6d9b72eb049)
![zoom-1](https://github.com/user-attachments/assets/c1cf8bcf-9e14-4b36-9e17-0449c9118b68)



The downside, of course, is that it requires forking the lib. That being said, maybe the maintainers would accept a new configuration option that would allow tweaking the scale before it is being applied:

```js
  focusOn: {
    region: 'XK',
    animate: true,
    transformScale: (scale) => Math.min(scale, Math.log(scale * 10)),
  },
```

<details>
<summary>Full git diff</summary>

```diff
diff --git a/packages/jsvectormap/src/js/core/setFocus.js b/packages/jsvectormap/src/js/core/setFocus.js
index 9abe6de..52ffe18 100644
--- a/packages/jsvectormap/src/js/core/setFocus.js
+++ b/packages/jsvectormap/src/js/core/setFocus.js
@@ -30,8 +30,11 @@ export default function setFocus(config = {}) {
       }
     })
 
+    const newScale = Math.min(this._width / (bbox.width), this._height / (bbox.height));
+    const adjustedScale = Math.min(newScale, Math.log(newScale * 10));
+    
     return this._setScale(
-      Math.min(this._width / bbox.width, this._height / bbox.height),
+      adjustedScale,
       -(bbox.x + bbox.width / 2),
       -(bbox.y + bbox.height / 2),
       true,
diff --git a/packages/playground/main.js b/packages/playground/main.js
index 553cb18..e0a7fda 100644
--- a/packages/playground/main.js
+++ b/packages/playground/main.js
@@ -1,5 +1,5 @@
 import jsVectorMap from 'jsvectormap'
-import 'jsvectormap/dist/maps/world-merc'
+import 'jsvectormap-maps/world-merc'
 import 'jsvectormap/dist/jsvectormap.css'
 
 const markers = [
@@ -23,19 +23,11 @@ const map = new jsVectorMap({
     initial: { fill: '#d1d5db' }
   },
 
-  // Labels
-  labels: {
-    markers: {
-      render: function(marker) {
-        return marker.name
-      },
-    },
+  focusOn: {
+    region: 'XK',
+    animate: true
   },
 
-  // Marker and label style
-  markers: markers,
-  markerStyle: { initial: { fill: '#66F' } },
-
   // Event handling
   onLoaded: function(map) {
     console.log(map)
diff --git a/packages/playground/package.json b/packages/playground/package.json
index 09c6130..78dcc52 100644
--- a/packages/playground/package.json
+++ b/packages/playground/package.json
@@ -6,7 +6,7 @@
   "scripts": {
     "library:dev": "pnpm run --filter=jsvectormap dev",
     "playground:dev": "vite",
-    "dev": "npm-run-all2 -p 'playground:dev' 'library:dev'",
+    "dev": "pnpm run \/:dev$\/",
     "build": "vite build",
     "preview": "vite preview"
   },
@@ -16,6 +16,6 @@
   },
   "dependencies": {
     "jsvectormap": "workspace:*",
-    "jsvectormap-maps": "*"
+    "jsvectormap-maps": "workspace:*"
   }
 }
```

</details>
